### PR TITLE
Store the 2FA backup codes hashed

### LINKED ISSUE #2047: Store 2FA backup codes hashed
To increase the 2FA security, we could hash the 2FA backup codes and store the hashes in the DB. This would imply that backup codes are only visible once and can only be regenerated when lost.

Related: #1852

### PR DESCRIPTION
**Pull request by bytehead**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2047 
| Docs PR or issue | n/a

Backup codes are hashed with the `PASSWORD_DEFAULT` algorithm now.

**TODO:**

- [x] Backend
- [x] Frontend
- [x] Migration script
- [x] Adjust tests

### CODE REVIEW

File: core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    private function updateBackupCodes(string $table): void
+    {
+        $rows = $this->getAffectedRowsForTable($table);
+
+        if (!\count($rows)) {
+            return;
+        }
+
```
Comment (ausi) [inline core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php]: This is not necessary I think because the foreach would just do nothing.

```suggestion
```

File: core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    public function shouldRun(): bool
+    {
+        $schemaManager = $this->connection->getSchemaManager();
+
+        if (!$schemaManager->tablesExist(['tl_user', 'tl_member'])) {
+            return false;
+        }
+
```
Comment (ausi) [inline core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php:41]: I think we should check if the `backupCodes` column exists to prevent SQL errors for old database structures.

File: core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php (hunk 1 above)
Comment (bytehead) [reply core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php]: Fixed in 5d072f6dd.

File: core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php (hunk 2 above)
Comment (bytehead) [reply core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php:41]: Added in 974fd974d.

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php [hunk 3]
```diff
@@ -42,12 +48,18 @@ public function invalidateBackupCode($user, string $code): void
...
-        $backupCodes = json_decode($user->backupCodes, true);
-        $key = array_search($code, $backupCodes, true);
+        $key = false;
+        $backupCodes = array_values(json_decode($user->backupCodes, true));
+
+        foreach ($backupCodes as $key => $backupCode) {
+            if (password_verify($code, $backupCode)) {
+                break;
+            }
+        }
```
Comment (ausi) [inline core-bundle/src/Security/TwoFactor/BackupCodeManager.php:59]: If no one of the backup codes matches, `$key` would be set to the last index and would be invalidated. This is probably not what we want?

**Review by ausi — CHANGES_REQUESTED** [review]

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php (hunk 3 above)
Comment (bytehead) [reply core-bundle/src/Security/TwoFactor/BackupCodeManager.php:59]: True story! 🙃

Comment (bytehead) [reply core-bundle/src/Security/TwoFactor/BackupCodeManager.php:59]: I've changed it in b1f496377.

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php [hunk 4]
```diff
@@ -42,15 +48,26 @@ public function invalidateBackupCode($user, string $code): void
...
-            $user->backupCodes = json_encode($backupCodes);
+        foreach ($backupCodes as $backupCode) {
+            if (password_verify($code, $backupCode)) {
+                $codeToInvalidate = $backupCode;
+                break;
+            }
         }
 
-        $user->save();
+        if (false !== $codeToInvalidate) {
```
Comment (leofeyer) [inline core-bundle/src/Security/TwoFactor/BackupCodeManager.php]: Early return here?

```php
if (false === $codeToInvalidate) {
    return;
}
```

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php [hunk 5]
```diff
@@ -42,15 +48,26 @@ public function invalidateBackupCode($user, string $code): void
...
+            $key = array_search($codeToInvalidate, $backupCodes, true);
+
+            if (false !== $key) {
```
Comment (leofeyer) [inline core-bundle/src/Security/TwoFactor/BackupCodeManager.php]: And here as well?

```php
if (false === $key) {
    return;
}
```

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php (hunk 4 above)
Comment (bytehead) [reply core-bundle/src/Security/TwoFactor/BackupCodeManager.php]: Fixed in 70a71ea17.

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php (hunk 5 above)
Comment (bytehead) [reply core-bundle/src/Security/TwoFactor/BackupCodeManager.php]: Fixed in 70a71ea17.

**Review by ausi — DISMISSED** [review]

**Review by aschempp — COMMENTED** [review]
Shouldn't we add a message to tell the user the codes are only shown once?

File: core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        return $this->createResult(true);
+    }
+
+    private function updateBackupCodes(string $table): void
+    {
+        $rows = $this->getAffectedRowsForTable($table);
+
+        foreach ($rows as $row) {
+            $backupCodes = json_decode($row['backupCodes'], true);
```
Comment (leofeyer) [inline core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php:68]: What if `json_decode()` fails? Do we have to handle this case?

File: core-bundle/src/Security/TwoFactor/BackupCodeManager.php [hunk 7]
```diff
@@ -61,7 +82,15 @@ public function generateBackupCodes(User $user): ?array
             $backupCodes[] = $this->generateCode();
         }
 
-        $user->backupCodes = json_encode($backupCodes);
+        $user->backupCodes = json_encode(
+            array_map(
+                static function ($backupCode) {
+                    return password_hash($backupCode, PASSWORD_DEFAULT);
```
Comment (leofeyer) [inline core-bundle/src/Security/TwoFactor/BackupCodeManager.php:88]: We could already hash the code in the `generateCode()` method, couldn't we? Then we do not have to traverse the array again.

Comment (ausi) [reply core-bundle/src/Security/TwoFactor/BackupCodeManager.php:88]: I don’t think so. The un-hashed `$backupCodes` get returned at the end of this method. AFAIK because we have to show the user the un-hashed codes (once).

File: core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php (hunk 6 above)
Comment (ausi) [reply core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php:68]: Should we ignore JSON errors here so that the migration can update as many rows as possible?
```suggestion
            $backupCodes = json_decode($row['backupCodes'], true);

            if (!\is_array($backupCodes)) {
                continue;
            }
```

Comment (leofeyer) [reply core-bundle/src/Migration/Version411/TwoFactorBackupCodesMigration.php:68]: Added in f9d23fe696eee867b88dd5359fe45ce20c3a5b85.

**Review by leofeyer — DISMISSED** [review]
The PR is working as expected, therefore I am approving the changes. ✅ 

However, there is an issue that bothers me: If you click on "Generate backup codes", the browser will send a POST request and the new backup codes are visible. Then, when you refresh the page, the POST request will be sent again and Contao will generate new backup codes and show them – even though the user has clicked on "Generate" and not on "Regenerate".

Is there a way to improve this? E.g. storing the backups codes in a flash message and reloading the page via GET?

**Comment by ausi** [comment]
> Then, when you refresh the page, the POST request will be sent again

Your browser should warn you when you do that.

> E.g. storing the backups codes in a flash message and reloading the page via GET?

The issue would still be happening for the back-button in this case. And more importantly: this would actually store the backup codes in plaintext on the server which we want to avoid with this PR.

> Is there a way to improve this?

We could try a JavaScript solution like `window.history.replaceState(null, null, window.location.href);` see <https://stackoverflow.com/a/45656609>

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 8]
```diff
@@ -1941,7 +1941,7 @@
         <source>Backup codes can be used to access your account if you have lost your authentication device and cannot generate two-factor codes anymore. Treat your backup codes with the same level of attention as your password! It is recommended to save them with a password manager such as Lastpass, 1Password or Keeper.</source>
       </trans-unit>
       <trans-unit id="MSC.twoFactorBackupCodesInfo">
-        <source>Put the codes in a safe spot. If you lose your authentication device and do not have backup codes, you cannot access your account anymore!</source>
+        <source>Make sure to copy the codes now. You will not be able to see them again!</source>
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/en/default.xlf:1944]: The original hint still seems valid to me, do we really need to remove it?

File: core-bundle/src/Resources/contao/modules/ModuleTwoFactor.php [hunk 9]
```diff
@@ -171,7 +171,7 @@ protected function disableTwoFactor(BackendUser $user, $return)
 	 *
 	 * @return array
 	 */
-	private function generateBackupCodes(BackendUser $user): ?array
+	private function generateBackupCodes(BackendUser $user)
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/modules/ModuleTwoFactor.php]: why this change?

File: core-bundle/src/Resources/contao/languages/en/default.xlf (hunk 8 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/default.xlf:1944]: The hint is already in the description above, so it makes sense to remove it here and use the info message to point out that the codes are only visible once.

File: core-bundle/src/Resources/contao/modules/ModuleTwoFactor.php (hunk 9 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/modules/ModuleTwoFactor.php]: CS (legacy code).

**Review by aschempp — DISMISSED** [review]

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.

**Comment by bytehead** [comment]
🎉
